### ISSUE TITLE
Build failure on mips with gcc 5.3 and botan 2.18.2

### ISSUE THREAD

**Issue by ffontaine** [randombit/botan#2845]
We have the following build failure on one mips toolchain since version 2.18.2 and https://github.com/randombit/botan/commit/219329f8e777af54d785ae7259f8be32a714b751:

```
src/lib/hash/sha3/sha3.cpp: In function 'std::tuple<long long unsigned int, long long unsigned int, long long unsigned int, long long unsigned int, long long unsigned int> Botan::{anonymous}::xor_CNs(const uint64_t*)':
src/lib/hash/sha3/sha3.cpp:41:42: error: converting to 'std::tuple<long long unsigned int, long long unsigned int, long long unsigned int, long long unsigned int, long long unsigned int>' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {long long unsigned int, long long unsigned int, long long unsigned int, long long unsigned int, long long unsigned int}; <template-parameter-2-2> = void; _Elements = {long long unsigned int, long long unsigned int, long long unsigned int, long long unsigned int, long long unsigned int}]'
       A[4] ^ A[9] ^ A[14] ^ A[19] ^ A[24]};
                                          ^
```

Build is fixed if I revert 219329f8e777af54d785ae7259f8be32a714b751. Do you have some advise? Should we add a dependency on gcc 6 or above?

Full build log: http://autobuild.buildroot.org/results/1aa/1aafc2bfdb82c6424327062e17f5745c15417526/build-end.log

**Comment by randombit** [comment]
Sorry about that. I did not intend to increase the minimum compiler version. I thought we had a GCC 4.4 CI build to ensure this did not happen, but possibly that was lost when we converted from Travis to GitHub Actions for CI.

As long as you are building only with GCC, reverting 219329f is just fine. I'll work on something to fix this for the next release.

**Comment by ffontaine** [comment]
Thanks for your feedback, I'll revert this commit on buildroot. It should be noted that 'm not 100% sure that this is only linked to the gcc version. This could also be used to this specific and very old toolchain.

**Comment by reneme** [comment]
GCC 4.4 seems to choke on the implicit call to the `std::tuple<>` constructor. By specifically adding it to the `return` statement, it seems to be happy: https://godbolt.org/z/5YnvvKTfT (see line 28).

**Comment by reneme** [comment]
See [here](https://github.com/randombit/botan/pull/2849) for a patch.

**Comment by ffontaine** [comment]
I confirm that #2849 fix the build failure on mips with gcc 5.3
